### PR TITLE
Deprecation warning

### PR DESCRIPTION
**Pull request by SwampDragons**
Update the fixer interface to include a method that returns the names of deprecated fields that the fixer fixes. 

Use that new method to generate a list of deprecated options.

Use that list of deprecated options when parsing to check whether unused keys are actually deprecated keys. If so, tell users how to update their templates using packer fix. 

This is a UX enhancement intended to make sure that users know about packer fix, and make the upgrade path to backwards-incompatible versions of Packer less painful. 

When reviewing, it'll be easier to step through the commits than view the whole thing at once; it is divided into three logical chunks.

### CODE REVIEW

File: fix/fixer_vsphere_network_storage.go [hunk 1]
```diff
@@ -8,6 +8,10 @@ import (
 // changes the disk_size, disk_thin_provisioned, and disk_eagerly_scrub into a storage adapter
 type FixerVSphereNetworkDisk struct{}
 
+func (FixerVSphereNetworkDisk) DeprecatedOptions() []string {
+	return []string{"network_card", "network", "networkCard"}
```
Comment (jhawk28) [inline fix/fixer_vsphere_network_storage.go:12]: missed the disk_size, disk_thin_provisioned, and disk_eagerly_scrub

Comment (nywilken) [reply fix/fixer_vsphere_network_storage.go:12]: Good catch @jhawk28

File: template/parse.go [hunk 2]
```diff
@@ -377,6 +377,8 @@ func Parse(r io.Reader) (*Template, error) {
 				rawTpl.Comments = append(rawTpl.Comments, comment)
 				continue
 			}
+			// Check for whether any of these keys are handled in a packer fix
```
Comment (nywilken) [inline template/parse.go:380]: Leftover comment?

File: fix/fixer_iso_md5.go [hunk 3]
```diff
@@ -8,6 +8,10 @@ import (
 // with the newer "iso_checksum" and "iso_checksum_type" within builders.
 type FixerISOMD5 struct{}
 
+func (FixerISOMD5) DeprecatedOptions() []string {
```
[the remark this answers is not in the document]
Comment (SwampDragons) [reply fix/fixer_iso_md5.go:11]: Order does matter. Off the top of my head I imagine this earlier fixer should run first, but I'll check into this.

[the remark this answers is not in the document]
Comment (SwampDragons) [reply fix/fixer_iso_md5.go:11]: Yep, rebased on master and the ordering works. if I use a template with the ancient "iso_md5" option it's first fixed into "iso_checksum" and "iso_checksum_type" and then fixed again so the final template uses the correct iso_checksum: md5:adfasdfadfaasdf format

File: cmd/generate-fixer-deprecations/main.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	absFilePath, err := filepath.Abs(fname)
+	if err != nil {
+		panic(err)
+	}
+	paths := strings.SplitAfter(absFilePath, "packer"+string(os.PathSeparator))
+	packerDir := paths[0]
+
+	// Load all deprecated options from all active fixers
+	allDeprecatedOpts := []string{}
+	for _, name := range fix.FixerOrder {
```
Comment (azr) [inline cmd/generate-fixer-deprecations/main.go:48]: Smart !!!

**Pull request merged**